fix shared library creation for aix (#278845)

http://bugs.python.org/issue941346 looks like an alternative approach
http://bugs.python.org/issue1542544 seems like a predecessor of the previous

--- configure.in
+++ configure.in
@@ -623,11 +623,16 @@
 	case $ac_sys_system in
 	AIX*)
 	   exp_extra="\"\""
+	   rtlflag=
 	   if test $ac_sys_release -ge 5 -o \
 		   $ac_sys_release -eq 4 -a `uname -r` -ge 2 ; then
-	       exp_extra="."
+	       exp_extra="'\$(LIBRARY)(\$(LDLIBRARY))'"
+	       if test "$GCC" = "yes"
+	       then rtlflag=' -Wl,-brtl'
+	       else rtlflag=' -brtl'
+	       fi
 	   fi
-	   LINKCC="\$(srcdir)/Modules/makexp_aix Modules/python.exp $exp_extra \$(LIBRARY); $LINKCC";;
+	   LINKCC="\$(srcdir)/Modules/makexp_aix Modules/python.exp $exp_extra \$(LIBRARY_OBJS); $LINKCC$rtlflag";;
 	Monterey64*)
 	   LINKCC="$LINKCC -L/usr/lib/ia64l64";;
 	QNX*)
@@ -693,6 +698,19 @@
 if test $enable_shared = "yes"; then
   AC_DEFINE(Py_ENABLE_SHARED, 1, [Defined if Python is built as a shared library.])
   case $ac_sys_system in
+    AIX*)
+	if test $ac_sys_release -ge 5 -o \
+		$ac_sys_release -eq 4 -a `uname -r` -ge 2
+	then
+	    INSTSONAME=
+	    LDLIBRARY='libpython$(VERSION).so.$(SOVERSION)'
+	    if test "$GCC" = "yes"
+	    then BLDLIBRARY='-Wl,-bI:Modules/python.exp -Wl,-blibpath:$(LIBDIR)'
+	    else BLDLIBRARY='-bI:Modules/python.exp -blibpath:$(LIBDIR)'
+	    fi
+	    RUNSHARED=LIBPATH=`pwd`:${LIBPATH}
+	fi
+        ;;
     BeOS*)
           LDLIBRARY='libpython$(VERSION).so'
           ;;
@@ -1623,8 +1644,24 @@
 then
 	case $ac_sys_system/$ac_sys_release in
 	AIX*)
-		BLDSHARED="\$(srcdir)/Modules/ld_so_aix \$(CC) -bI:Modules/python.exp"
-		LDSHARED="\$(BINLIBDEST)/config/ld_so_aix \$(CC) -bI:\$(BINLIBDEST)/config/python.exp"
+	  if test "$GCC" = "yes"
+	  then
+		sharedflag=' -shared'
+		if test $ac_sys_release -ge 5 -o \
+			$ac_sys_release -eq 4 -a `uname -r` -ge 2 ; then
+		  sharedflag="$sharedflag -Wl,-G -Wl,-bernotok"
+		fi
+		BLDSHARED="\$(srcdir)/Modules/ld_so_aix \$(CC)$sharedflag -bI:Modules/python.exp"
+		LDSHARED="\$(BINLIBDEST)/config/ld_so_aix \$(CC)$sharedflag -bI:\$(BINLIBDEST)/config/python.exp"
+	  else
+		sharedflag=
+		if test $ac_sys_release -ge 5 -o \
+			$ac_sys_release -eq 4 -a `uname -r` -ge 2 ; then
+		  sharedflag="$sharedflag -G -bernotok"
+		fi
+		BLDSHARED="\$(srcdir)/Modules/ld_so_aix \$(CC)$sharedflag -bI:Modules/python.exp"
+		LDSHARED="\$(BINLIBDEST)/config/ld_so_aix \$(CC)$sharedflag -bI:\$(BINLIBDEST)/config/python.exp"
+	  fi
 		;;
 	BeOS*)
 		BLDSHARED="\$(srcdir)/Modules/ld_so_beos $LDLIBRARY"
--- Makefile.pre.in.orig	2009-07-24 17:15:25 +0200
+++ Makefile.pre.in	2009-07-24 17:19:03 +0200
@@ -25,6 +25,7 @@
 
 # === Variables set by configure
 VERSION=	@VERSION@
+SOVERSION=	@SOVERSION@
 srcdir=		@srcdir@
 VPATH=		@srcdir@
 
@@ -428,6 +429,15 @@
 libpython$(VERSION).dylib: $(LIBRARY_OBJS)
 	$(LINKCC) -dynamiclib -install_name $(LIBDIR)/$@ -current_version $(VERSION) -compatibility_version 2.6 -o $@ $(LIBRARY_OBJS) $(SHLIBS) $(LIBC) $(LIBM)
 
+# AIX likes the shared library inside the archive library.
+libpython$(VERSION).so.$(SOVERSION): $(LIBRARY)
+	$(LINKCC) -shared -Wl,-G -Wl,-bernotok $(LDFLAGS) $(LINKFORSHARED) -o $@ $(LIBRARY_OBJS) $(MODLIBS) $(SHLIBS) $(LIBC) $(LIBM) $(LDLAST)
+	touch -r $(LIBRARY) $(LIBRARY).stamp
+	-rm -f $(LIBRARY)
+	$(AR) cr $(LIBRARY) $@
+	$(RANLIB) $(LIBRARY)
+	touch -r $(LIBRARY).stamp $(LIBRARY)
+	rm $(LIBRARY).stamp
 
 # This rule is here for OPENSTEP/Rhapsody/MacOSX. It builds a temporary
 # minimal framework (not including the Lib directory and such) in the current
@@ -783,7 +793,7 @@
 	if test -f $(LDLIBRARY); then \
 		if test -n "$(DLLLIBRARY)" ; then \
 			$(INSTALL_SHARED) $(DLLLIBRARY) $(DESTDIR)$(BINDIR); \
-		else \
+		elif test -n "$(INSTSONAME)" ; then \
 			$(INSTALL_SHARED) $(LDLIBRARY) $(DESTDIR)$(LIBDIR)/$(INSTSONAME); \
 			if test $(LDLIBRARY) != $(INSTSONAME); then \
 				(cd $(DESTDIR)$(LIBDIR); $(LN) -sf $(INSTSONAME) $(LDLIBRARY)) \
